app/questionnaire/path_finder.py

```diff
@@ -134,35 +135,56 @@ def _build_routing_path_block_ids(self, blocks, current_location):
-                self.metadata,
-                self.answer_store,
-                self.list_store,
-                current_location=this_location,
-                routing_path_block_ids=routing_path_block_ids,
-            )
+        if any("goto" in rule for rule in routing_rules):
+            for rule in filter(is_goto_rule, routing_rules):
+                should_goto = evaluate_goto(
+                    rule["goto"],
+                    self.schema,
+                    self.metadata,
+                    self.answer_store,
+                    self.list_store,
+                    current_location=this_location,
+                    routing_path_block_ids=routing_path_block_ids,
+                )
 
-            if should_goto:
-                if rule["goto"].get("section") == "End":
-                    return None
+                if should_goto:
+                    if rule["goto"].get("section") == "End":
+                        return None
 
-                next_block_id = self._get_next_block_id(rule)
-                next_block_index = PathFinder._block_index_for_block_id(
-                    blocks, next_block_id
-                )
-                next_precedes_current = (
-                    next_block_index is not None and next_block_index < block_index
-                )
+                    next_block_id = self._get_next_block_id(rule)
+                    next_block_index = PathFinder._block_index_for_block_id(
+                        blocks, next_block_id
+                    )
+                    next_precedes_current = (
+                        next_block_index is not None and next_block_index < block_index
+                    )
 
-                if next_precedes_current:
-                    self._remove_rule_answers(rule["goto"], this_location)
-                    routing_path_block_ids.append(next_block_id)
-                    return None
+                    if next_precedes_current:
+                        self._remove_rule_answers(rule["goto"], this_location)
+                        routing_path_block_ids.append(next_block_id)
+                        return None
+
+                    return next_block_index
+        else:
+            for rule in routing_rules:
+                should_goto = should_goto_new(
+                    rule,
+                    self.schema,
+                    self.answer_store,
+                    self.list_store,
+                    self.metadata,
+                    this_location,
+                    routing_path_block_ids,
+                )
+                if should_goto:
+                    if rule.get("section") == "End":
+                        return None
+                    next_block_id = rule["block"]
+                    next_block_index = PathFinder._block_index_for_block_id(
+                        blocks, next_block_id
+                    )
 
-                return next_block_index
+                    return next_block_index
```

Comment:
The logic within `should_goto` needs to be the same for both. I think we can handle both in the same loop with something like:
```python
        for rule in routing_rules:
            if "goto" in rule:
                rule = rule["goto"]
                should_goto = evaluate_goto(
                    rule,
                    self.schema,
                    self.metadata,
                    self.answer_store,
                    self.list_store,
                    current_location=this_location,
                    routing_path_block_ids=routing_path_block_ids,
                )
            else:
                should_goto = should_goto_new(
                    rule,
                    self.schema,
                    self.answer_store,
                    self.list_store,
                    self.metadata,
                    this_location,
                    routing_path_block_ids,
                )

            if should_goto:
                if rule.get("section") == "End":
                    return None

                next_block_id = self._get_next_block_id(rule)
                next_block_index = PathFinder._block_index_for_block_id(
                    blocks, next_block_id
                )
                next_precedes_current = (
                    next_block_index is not None and next_block_index < block_index
                )

                if next_precedes_current:
                    self._remove_rule_answers(rule, this_location)
                    routing_path_block_ids.append(next_block_id)
                    return None

                return next_block_index
```

Reply:
All this above added now so it's the same loop and same code reused. Thanks!